### PR TITLE
Implement a rendering pipeline with middleware (adds support for Brotli and Gzip, cleans up minification)

### PR DESCRIPTION
**Pull request by EvanCarroll**
I've updated this architecture to support a rendering pipeline which everything goes through.

* Minification is now a middleware.
* Livereload is now a middleware.

At the end of every pipeline we call `write()`. It's no longer called as IO call at the end of a function (which only know it was writing an `index.html`, and thus would minify `.txt`). This works and passes all tests,

* livereload knows if it's in a pipeline generated from an HTML
* minification knows if its in a pipeline generating HTML

With this infrastructure, this patch

* adds compression closing #1684
* and encryption #2993

AI summary of patch..

  Core Architecture

```
  ┌─────────────┐
  │  Renderable │ (trait)
  └──────┬──────┘
         │
         ├─ Page
         ├─ Section
         ├─ Taxonomy
         ├─ Pager
         ├─ Feed
         ├─ Sitemap
         └─ ... (11 types)
         │
         ▼
  ┌──────────────────┐
  │    Pipeline      │
  ├──────────────────┤
  │ 1. Render        │◄── Tera templates
  │ 2. Metadata      │
  │ 3. Middleware    │◄── Pluggable chain
  │ 4. Process       │
  │ 5. Write         │
  └──────────────────┘
         │
         ├─ LiveReloadMiddleware
         ├─ MinifyMiddleware
         └─ [future: compression, encryption]
         │
         ▼
  ┌──────────────────┐
  │  ProcessedContent│
  └──────────────────┘
         │
         ▼
  ┌──────────────────┐
  │  ContentWriter   │
  ├──────────────────┤
  │ BuildMode::Disk  │─── File system
  │ BuildMode::Memory│─── In-memory cache
  │ BuildMode::Both  │─── Both
  └──────────────────┘
```

  Key Components

  1. Renderable Trait (renderable.rs)

  - Purpose: Unified interface for all content types
  - Methods:
    - render() - produces HTML/XML/JSON from templates
    - metadata() - provides routing & content type info
  - Implementations: 11 types (Page, Section, Taxonomy, Feed, Sitemap, robots.txt, 404, Pager, Redirect, etc.)

  2. Pipeline (pipeline.rs)

  - Purpose: Orchestrates the 5-stage rendering process
  - Stages:
    a. Render: Template → String (via Tera)
    b. Metadata: Extract path, content type, permalink
    c. Context: Wrap in MiddlewareContext
    d. Middleware Chain: Sequential transformations
    e. Return: ProcessedContent ready to write

  3. Middleware Trait (middleware.rs)

  - Purpose: Extensible content transformation
  - Interface: process(&self, ctx: &mut MiddlewareContext)
  - Context includes:
    - content: String - the HTML/XML/JSON
    - metadata: ContentMetadata - path, template, content type
    - config: Arc<Config> - site configuration

  4. Concrete Middleware

  - LiveReloadMiddleware (middleware/livereload.rs):
    - Injects <script src="/livereload.js"> before </body>
    - Only processes ContentType::Html
  - MinifyMiddleware (middleware/minify.rs):
    - Removes whitespace from HTML
    - Conditional on config.minify_html
    - Only processes ContentType::Html

  5. ContentWriter (writer.rs)

  - Purpose: Handles output destination
  - BuildMode:
    - Disk: Write to filesystem (public/ directory)
    - Memory: Store in SITE_CONTENT cache (for live reload)
    - Both: Do both operations

  Design Benefits

  1. Separation of Concerns: Rendering, processing, and writing are isolated
  2. Type Safety: ContentType enum ensures correct middleware application
  3. Extensibility: Add new middleware without touching core logic
  4. Testability: Each component has unit tests
  5. Future-Ready: Designed for compression & encryption (per commit message)
  6. Unified Flow: All 11 content types go through the same pipeline

  Before vs After

  Before: Rendering logic scattered, minification/livereload handled per-type

  After: Single pipeline with pluggable middleware - every rendered item flows through the same chain

  This architecture mirrors common web frameworks (Express.js middleware, ASP.NET pipeline) and provides a clean foundation for adding post-processing features like
  compression or encryption.
  
 

----

* `write_content` isn't smart enough to do what we need.
* this is a cleaner architecture.
* Now pages with txt templates are not minified.
* Add test for text template whitespace preservation with minification
  * Test verifies that non-HTML templates (e.g., .txt) preserve whitespace
  * Created text_content.txt template with multiple spaces, tabs, and newlines
  * Created test page using the text template
  * Test passes with the minification fix (ed36c1e6) and fails without it
  * Updated page count assertions due to new test page

Addresses #3006

### CODE REVIEW

File: components/site/src/lib.rs [hunk 1]
```diff
@@ -889,7 +897,11 @@ impl Site {
             None => "index.html",
         };
         let content = render_redirect_template(permalink, &self.tera)?;
-        self.write_content(&split, page_name, content)?;
+
+        // Aliases are always HTML redirects - minify if enabled
+        let final_content = self.maybe_minify(content)?;
```
Comment (Keats) [inline components/site/src/lib.rs]: Hm that's not great, we're just having to remember to call that everywhere we're writing HTML instead of once. We should instead be able to see whether the output is HTML or not and not generate an index.html for those like we do for sitemaps.

**Review by Keats — CHANGES_REQUESTED** [review]

File: components/site/src/lib.rs (hunk 1 above)
Comment (EvanCarroll) [reply components/site/src/lib.rs]: That specific part of the file you've highlighted is how the code currently works. `maybe_minify` does 

```
    fn maybe_minify(&self, content: String) -> Result<String> {
        if self.config.minify_html { minify::html(content) } else { Ok(content) }
    }
```

That is to say, it's just a conditional to check whether or not the config option is present at all. We can roll that into the routine itself if you find it confusing.

----

> Hm that's not great, we're just having to remember to call that everywhere we're writing HTML instead of once. We should instead be able to see whether the output is HTML or not and not generate an index.html for those like we do for sitemaps.

Fixing that would require re-architecting this. We currently have,

* The template,
* The filename you're writing.

And

* `write_content`, is the one that that writes files. it has no idea of the template.
* A lot of `render_` where each one call `write_content` with the content it's writing.

----

Problem, 

* If anything can `write_content`
* And, `write_content` isn't template aware and only knows of content and a destination file location,
* How can I tell whether or not minification is needed?

Currently you have to do that at rendering step.

----

So ultimately to do this without having to have the logic in everything that can produce html, we would need to create something between the renderer and the IO layer.

If you want a more drastic rewrite, I'll create you a middleware rendering pipeline that can solve this problem. Where anywhere middleware can answer the call at the end. But this requires blocking access to `write_content` and having the render's return the mime/type that that they're rendering to the next step in the pipeline.

This could also be used to solve, #1684

Comment (EvanCarroll) [reply components/site/src/lib.rs]: I'm going to close this because I rewrote all this and a lot more.

**Comment by EvanCarroll** [comment]
@Keats if you determine this architecture isn't something you want to move towards, please tell me so I can make adjustments or quit closing out issues.

**Comment by ccjmne** [comment]
Seems really cool, but the current MR is **a lot** for a human to digest.
Do you think you could keep the encryption middleware in another PR for the time being? Though maybe @Keats can review it like that.

**Comment by EvanCarroll** [comment]
> Seems really cool, but the current MR is **a lot** for a human to digest. Do you think you could keep the encryption middleware in another PR for the time being? Though maybe @Keats can review it like that.

Fair and certainly doable. Because the architecture isn't cleanly in one patch it makes this patch even more complex. I tweak it a few times. If this is needed. what I can do is rewrite the the patches entirely to sink the commits separately. It's a waste of time though if that's not needed.

Actually if the middlewares themselves had their own config parser, they would be entirely self-contained and this would be trivial..

The next patch that I want to do pulls out markdown rendering putting that into the middleware too. And closes out #2930, and #2420 and #2790. This would make adding first-class Tyrst the same way trivial.

**Comment by Keats** [comment]
It is indeed way too big to review. A few other things:

- The issue for secret content is closed so for now it shouldn't be implemented
- Same-ish for pre-compression, I don't think it's super important
- I think it's too complex overall? All we need to have is a way to differentiate HTML and other content and do different operations when wrtiting the content.

**Comment by EvanCarroll** [comment]
I'll get back this and rebase the patch to make it easier to process. I got one more project I have to get off my plate. The architectural changes, should make all the middle wares one file (or relatively small) so you can pick whatever middleware you want.

**Comment by Keats** [comment]
The `next` branch now has some kind of queue processing.
